### PR TITLE
Add instructions for writing unit tests

### PR DESCRIPTION
**Pull request by weiji14**
Instructions for activity on final day for learning about basics of writing unit tests using `pytest`.

Preview slides at https://hackmd.io/@pygmt/eswn-2021-writing-unit-tests

TODO:
- [x] Check on where to fit this in the schedule?
- [x] I might add a few more PyGMT-specific slides based on https://www.pygmt.org/v0.4.1/contributing.html#testing-your-code, depending on how deep we want to go.

### CODE REVIEW

File: activities/writing-unit-tests.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+```markdown
+projectname/
+  ├──src/
+  │   └──calc.py (contains 'add' function)
+  └──tests/
+      └──test_calc.py (contains 'test_add' function)
+```
+
+The 'test_addition.py' file would contain:
```
Comment (maxrjones) [inline activities/writing-unit-tests.md]:
```suggestion
The 'test_calc.py' file would contain:
```

**Review by maxrjones — COMMENTED** [review]
This is great, thank you!

Do you have a preference on where it goes in the schedule? I'm fine adding it anywhere after the second exercise.

**Comment by weiji14** [comment]
> Do you have a preference on where it goes in the schedule? I'm fine adding it anywhere after the second exercise.

Sorry, could you be more specific on which second exercise you mean?

**Comment by maxrjones** [comment]
> > Do you have a preference on where it goes in the schedule? I'm fine adding it anywhere after the second exercise.
> 
> Sorry, could you be more specific on which second exercise you mean?

Yes, referring to: First issues - making a pull request (local workflow)

**Comment by weiji14** [comment]
> > > Do you have a preference on where it goes in the schedule? I'm fine adding it anywhere after the second exercise.
> > 
> > Sorry, could you be more specific on which second exercise you mean?
> 
> Yes, referring to: First issues - making a pull request (local workflow)

Owh, I was thinking sometime tomorrow (19 Aug)... Like in the 20:00-20:15 (UTC) slot maybe? Let's discuss later.

**Comment by maxrjones** [comment]
> > > > Do you have a preference on where it goes in the schedule? I'm fine adding it anywhere after the second exercise.
> > > 
> > > 
> > > Sorry, could you be more specific on which second exercise you mean?
> > 
> > 
> > Yes, referring to: First issues - making a pull request (local workflow)
> 
> Owh, I was thinking sometime tomorrow (19 Aug)... Like in the 20:00-20:15 (UTC) slot maybe? Let's discuss later.

Tomorrow is great. We can set a time after today's workshop.

**Review by maxrjones — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -83,13 +83,13 @@ Participants will receive zoom invitations via email.
 
 | **Theme** | **Duration** | **UTC Time** |
 |-----------|--------------|--------------|
-| Second issue - break-out group discussion | 15 min | 20:00 - 20:15 |
-| Second issue - group reports | 10 min | 20:15 - 20:25 |
+| [Writing unit tests](activities/writing-unit-tests.md) | 15 min | 20:00 - 20:15 |
+| Second issue - break-out group discussion / group reports | 10 min | 20:15 - 20:25 |
 | Second issue - local development | 25 min | 20:25 - 20:50 |
 | Break | 10 min | 20:50 - 21:00 |
 | Second issue - submitting through issues/pull requests | 50 min |21:00 - 21:50 |
```
Comment (weiji14) [inline README.md]: Let's discuss here. I was thinking of moving the group discussion / group reports to be 20:15-20:50, and then do the local development after 21:00. May need to squeeze some of the activities after 21:00 to be shorter (did you have any material prepared for those?).

Comment (maxrjones) [reply README.md]: That would be fine with me. I don't have any material prepared. My plan was for people to chose issues or other contributions to work on with us available to help (sort of like a mini-sprint). Does that sound good to you or would you prefer more structured material?

Comment (weiji14) [reply README.md]: Yep, a mini-sprint sounds good. How about we post on the Slack channel to find out what issues they would like to work on (i.e. with the eswn-workshop label, or any other they fancy), so we can assign issues to them.

Meanwhile, I'll try to squeeze some sections in the schedule together and you can comment on how it looks.

Comment (maxrjones) [reply README.md]: Sounds great!

File: README.md [hunk 3]
```diff
@@ -83,13 +83,13 @@ Participants will receive zoom invitations via email.
...
+| Second issue - break-out group discussion | 15 min | 20:20 - 20:35 |
+| Second issue - group reports | 15 min | 20:35 - 20:50 |
 | Break | 10 min | 20:50 - 21:00 |
-| Second issue - submitting through issues/pull requests | 50 min |21:00 - 21:50 |
+| Second issue - local development to work through issues/pull requests | 50 min | 21:00 - 21:50 |
 | Break | 10 min | 21:50 - 22:00 |
-| Second issue - incorporating suggestions | 50 min | 22:00 - 22:50 |
+| Second issue - incorporating suggestions from reviewers | 50 min | 22:00 - 22:50 |
 | Break | 10 min | 22:50 - 23:00 |
 | Helping maintain software | 45 min | 23:00 - 23:45 |
```
Comment (maxrjones) [inline README.md]: The helping maintain software section might be a bit long. My original plan was to lump testing into this section, so I think we should shorten it now.
```suggestion
| Second issue -  continue working on contributions | 30 min | 23:00 - 23:30 |
| Maintaining software | 10 min | 23:30 - 23:23:40 |
| Resources for further learning | 5 min | 23:40 - 23:45 |
```

File: activities/writing-unit-tests.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+## Where do the tests go?
+
+In a Python project using [`pytest`](https://docs.pytest.org/en/6.2.x/), we usually structure our code in folders like so:
+
+```markdown
+projectname/
+  ├──src/
+  │   └──calc.py (contains 'add' function)
+  └──tests/
+      └──test_calc.py (contains 'test_add' function)
```
Comment (maxrjones) [inline activities/writing-unit-tests.md]:
```suggestion
projectname/
  ├──__init__.py
  ├──src/
  │   └──calc.py (contains 'add' function)
  └──tests/
      ├──test_calc.py (contains 'test_add' function)
      └──__init__.py
```
Not sure if I formatted the tree properly, but `__init__.py` is needed for the example to work.

**Review by maxrjones — APPROVED** [review]

**Pull request merged**